### PR TITLE
[RSDK-1038] Adding use_live_data parameter to SLAM

### PR DESCRIPTION
**Pull request by jeremyhyde-viam**
Adds a new argument, 'use_live_data' to SLAM, defined in the [SLAM usability scope doc](https://docs.google.com/document/d/1RsT-c0QOtkMKa-rwUGY0-emUmKIRSaO3mzT1v6MtFjk/edit?usp=sharing), that is responsible for indicating if newly generated data or previously generated data should be used during this run. This is replacing the previous offlineFlag that was auto populated based on the presence or lack of sensors. 

This PR is made in conjunction with the PR in RDK that adds use_live_data both to the list of arguments passed to SLAM but also the configs in testing.

In order to ensure backwards compatibility currently in orbslam we check if the use_live_data argument exists and if not, we use the old method of the sensor list to determine online/offline mode for cartographer we currently over write the offline_flag with the state of the sensor list. This means that currently there is no way to run in offline mode when a sensor is given and no way to run in online mode when no sensor is given. These features will be added back in the PR to remove backwards compatibility

JIRA Ticket: [RSDK-1038](https://viam.atlassian.net/browse/RSDK-1038)
Associated PR: [#1729](https://github.com/viamrobotics/rdk/pull/1729)

Remove backwards compatibility JIRA Ticket:  [RSDK-1625](https://viam.atlassian.net/browse/RSDK-1625)

[RSDK-1038]: https://viam.atlassian.net/browse/RSDK-1038?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: slam-libraries/viam-cartographer/src/slam_service/config.cc
Comment (jeremyhyde-viam) [inline slam-libraries/viam-cartographer/src/slam_service/config.cc:78]: Do we want to add in an error if online mode is given with an empty list of sensors (and should this also be initially checked in RDK?)

File: slam-libraries/viam-cartographer/src/slam_service/config.cc
Comment (jeremyhyde-viam) [inline slam-libraries/viam-cartographer/src/slam_service/config.cc:35]: Default value for flag is currently set to false, although since it is mandatory this should never affect operation)

File: slam-libraries/viam-cartographer/src/slam_service/config.cc
Comment (JohnN193) [reply slam-libraries/viam-cartographer/src/slam_service/config.cc:78]: i think only having the check in rdk is fine, as the slam algorithms should never run into that test case

Comment (jeremyhyde-viam) [reply slam-libraries/viam-cartographer/src/slam_service/config.cc:78]: I agree and have left a similar comment to that affect in the RDK PR as the issue of erroring out was not covered in the scope doc

**Review by JohnN193 — APPROVED** [review]
LGTM

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc:286]: [minor] We should replace `slamService.offline_flag` with `slamService.use_live_data` and test all four cases (`use_live_data` vs. `sensors`).

**Review by kkufieta — CHANGES_REQUESTED** [review]
First pass.

File: slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc
Comment (jeremyhyde-viam) [reply slam-libraries/viam-orb-slam3/orbslam_server_v1_test.cc]: all tested!

File: slam-libraries/viam-cartographer/src/slam_service/config.cc
Comment (tessavitabile) [reply slam-libraries/viam-cartographer/src/slam_service/config.cc]: We've been a little inconsistent in what validation occurs in orb_grpc_server/carto_grpc_server. I would say that parameter validation *must* occur rdk. In some cases, we've additionally chosen to include validation in orb_grpc_server/carto_grpc_server, primarily when the program cannot continue without it (e.g. a missing required parameter). My suggestion is to avoid logical validation in orb_grpc_server/carto_grpc_server (e.g. a disallowed combination of parameters), since this duplicates validation in rdk. However, I don't feel so strongly about this, particularly since we will eventually combine the slam service with orb_grpc_server/carto_grpc_server.

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc
Comment (kkufieta) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:474]: [nit] This is easier to read:

```go
 if (use_live_data) {
      return GetNextDataFileOnline();
  }
  return GetNextDataFileOffline();
```

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc
Comment (kkufieta) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:503]: [nit] I think you can remove the braces? `(!use_live_data && ...`

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc
Comment (kkufieta) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:518]: [nit] I think you can remove the braces? `(!use_live_data && ...`

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:785]: [nit] I think you can remove the braces? `!slamService.camera_name.empty()`

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:806]: [nit] I think you can remove the braces? `!slamService.use_live_data`

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.cc
Comment (kkufieta) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.cc:814]: [nit] I think you can remove the braces? `!slamService.use_live_data`

**Review by kkufieta — APPROVED** [review]
This looks great! LGTM mod nits

**Comment by tessavitabile** [comment]
I can't remove myself from reviewers, but I'm not planning to do a full review, so no need to wait for my sign-off.

**Pull request merged**
